[CAS] Add an option to validate CASID in CASDB

cachemeifyoucan opened the pull request:
Add a debug option that validate a CASID in the CASDB by loading the
object then rehash its content to make sure it matches the hash value
from the CASID.

cachemeifyoucan commented:
@swift-ci please test

benlangmuir commented on llvm/lib/CAS/OnDiskCAS.cpp:
```diff
@@ -325,6 +325,7 @@ class InternalRefVector {
     FullRefs.reserve(SmallRefs.size() + 1);
     for (InternalRef4B Small : SmallRefs)
       FullRefs.push_back(Small);
+    FullRefs.push_back(Ref);
```
I merged this change, so should probably remove it here to avoid accidentally duplicating it.

cachemeifyoucan replied:
I thought I cleaned this up and all the some of the comments below but apparently I edited the headers in a different directory. Will fix them.

benlangmuir commented on llvm/tools/llvm-cas/llvm-cas.cpp:
```diff
@@ -7,11 +7,13 @@
 #include "llvm/CAS/CachingOnDiskFileSystem.h"
 #include "llvm/CAS/HierarchicalTreeBuilder.h"
 #include "llvm/CAS/Utils.h"
+#include "llvm/Support/BLAKE3.h"
```
Is this (and the BuiltinObjectHasher.h above) needed?

benlangmuir commented on llvm/include/llvm/CAS/CASDB.h:
```diff
@@ -248,6 +248,8 @@ class CASDB : public CASIDContext {
   template <class ProxyT, class HandleT>
   Expected<ProxyT> loadObjectProxy(Expected<HandleT> H);
 
+  virtual Error validateID(const CASID &ID) = 0;
```
How about changing this to `validateObject`? I think validating an "ID" is unclear that it's validating the whole object not just the ID itself.

cachemeifyoucan committed:
[CAS] Add an option to validate CASID in CASDB

Add a debug option that validate a CASID in the CASDB by loading the
object then rehash its content to make sure it matches the hash value
from the CASID.

cachemeifyoucan commented:
@swift-ci please test
